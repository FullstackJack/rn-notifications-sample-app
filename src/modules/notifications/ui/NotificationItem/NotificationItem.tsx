import React, {memo} from 'react';
import {StyleSheet, Text, View} from 'react-native';
import {useDispatch} from 'react-redux';
import {formatDistance} from 'date-fns';
import {INotification} from '../../types';
import {NotificationText} from '../NotificationText';
import {FlatButton} from '../../../../common/ui/FlatButton';
import {OutlineButton} from '../../../../common/ui/OutlineButton';
import {markAsRead} from '../../data/notificationsSlice';
import {ACTION_FRIEND_REQUEST} from '../../data/constants';

interface INotificationViewProps extends INotification {}

const NotificationItem = memo(function NotificationItem(
  props: INotificationViewProps,
) {
  const {action, isRead, id: notificationId} = props;
  const dispatch = useDispatch();
  const markReadAction = (id: string) => () => dispatch(markAsRead({id}));
  const opacity = isRead ? 0.5 : 1;
  const timeago = formatDistance(
    Date.now() - new Date().getTimezoneOffset() * 60000,
    props.createdAt,
  );

  return (
    <View style={styles.container}>
      <View style={styles.avatar} />
      <View style={{...styles.content, opacity}}>
        <NotificationText {...props} />
        <Text style={styles.timeago}>{timeago}</Text>
        {!isRead &&
          (action === ACTION_FRIEND_REQUEST ? (
            <View style={styles.buttonGroup}>
              <FlatButton
                testID="accept-button"
                text="Accept"
                onPress={markReadAction(notificationId)}
              />
              <OutlineButton
                testID="decline-button"
                text="Decline"
                onPress={markReadAction(notificationId)}
              />
            </View>
          ) : (
            <FlatButton
              text="Mark Read"
              onPress={markReadAction(notificationId)}
            />
          ))}
      </View>
    </View>
  );
});

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    paddingVertical: 8,
    paddingHorizontal: 8,
    columnGap: 10,
    backgroundColor: '#273246',
  },
  avatar: {
    width: 35,
    height: 35,
    backgroundColor: 'purple',
    borderRadius: 8,
  },
  timeago: {
    color: '#ccc',
  },
  content: {
    rowGap: 4,
  },
  buttonGroup: {flexDirection: 'row', columnGap: 8},
});

export default NotificationItem;
