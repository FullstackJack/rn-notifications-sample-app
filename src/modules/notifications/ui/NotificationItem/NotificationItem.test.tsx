import React from 'react';
import {expect, test} from '@jest/globals';
import {screen} from '@testing-library/react-native';
import NotificationItem from './NotificationItem';
import {renderWithProviders} from '../../../../common/utils/test-utils';

test('renders correctly', async () => {
  renderWithProviders(
    <NotificationItem
      id="n0"
      action="mention"
      createdAt={Date.UTC(2023, 0, 1, 0).valueOf()}
      channel={{
        name: 'Client',
        id: 'ch123',
      }}
      community={{
        name: 'Root',
        id: 'c123',
      }}
      isRead={false}
      user={{
        name: 'Vinny',
        id: 'u123',
      }}
    />,
  );

  expect(screen.toJSON()).toMatchSnapshot();
});
