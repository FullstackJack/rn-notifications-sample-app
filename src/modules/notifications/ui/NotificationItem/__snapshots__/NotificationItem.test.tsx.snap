// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#273246",
      "columnGap": 10,
      "flexDirection": "row",
      "paddingHorizontal": 8,
      "paddingVertical": 8,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "purple",
        "borderRadius": 8,
        "height": 35,
        "width": 35,
      }
    }
  />
  <View
    style={
      {
        "opacity": 1,
        "rowGap": 4,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "flexWrap": "wrap",
        }
      }
    >
      <Text
        style={
          {
            "backgroundColor": "#39465a",
            "color": "white",
            "padding": 2,
          }
        }
      >
        @
        Vinny
      </Text>
      <Text
        style={
          {
            "color": "#eee",
            "padding": 2,
          }
        }
      >
         mentioned you in
      </Text>
      <Text
        style={
          {
            "color": "#eee",
            "padding": 2,
          }
        }
      >
        #
        Client
        .
        Root
      </Text>
    </View>
    <Text
      style={
        {
          "color": "#ccc",
        }
      }
    >
      1 hour ago
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "alignSelf": "flex-start",
            "backgroundColor": "white",
            "borderColor": "transparent",
            "borderRadius": 50,
            "borderStyle": "solid",
            "borderWidth": 1,
            "justifyContent": "center",
            "paddingHorizontal": 16,
            "paddingVertical": 8,
            "width": "auto",
          },
          undefined,
          {
            "backgroundColor": "white",
          },
        ]
      }
    >
      <Text>
        Mark Read
      </Text>
    </View>
  </View>
</View>
`;
